src/metpy/calc/tools.py

```diff
@@ -950,6 +980,219 @@ def wrapper(f, **kwargs):
+        bound_args.apply_defaults()
+        scale_lat = latitude = bound_args.arguments.pop('latitude')
+        scale_lon = longitude = bound_args.arguments.pop('longitude')
+        crs = bound_args.arguments.pop('crs')
+
+        # Choose the first DataArray argument to act as grid prototype
+        grid_prototype = next(dataarray_arguments(bound_args), None)
+
+        # Fill in x_dim/y_dim
+        if (
+            grid_prototype is not None
+            and 'x_dim' in bound_args.arguments
+            and 'y_dim' in bound_args.arguments
+        ):
+            try:
+                bound_args.arguments['x_dim'] = grid_prototype.metpy.find_axis_number('x')
+                bound_args.arguments['y_dim'] = grid_prototype.metpy.find_axis_number('y')
+            except AttributeError:
+                # If axis number not found, fall back to default but warn.
+                warnings.warn('Horizontal dimension numbers not found. Defaulting to '
+                              '(..., Y, X) order.')
+
+        # Fill in vertical_dim
+        if (
+            grid_prototype is not None
+            and 'vertical_dim' in bound_args.arguments
+        ):
+            try:
+                bound_args.arguments['vertical_dim'] = (
+                    grid_prototype.metpy.find_axis_number('vertical')
+                )
+            except AttributeError:
+                # If axis number not found, fall back to default but warn.
+                warnings.warn(
+                    'Vertical dimension number not found. Defaulting to (..., Z, Y, X) order.'
+                )
+
+        # Fill in dz
+        if (
+            grid_prototype is not None
+            and 'dz' in bound_args.arguments
+            and bound_args.arguments['dz'] is None
+        ):
+            try:
+                vertical_coord = grid_prototype.metpy.vertical
+                bound_args.arguments['dz'] = np.diff(vertical_coord.metpy.unit_array)
+            except (AttributeError, ValueError):
+                # Skip, since this only comes up in advection, where dz is optional (may not
+                # need vertical at all)
+                pass
+
+        # Fill in dx and dy
+        if (
+            'dx' in bound_args.arguments and bound_args.arguments['dx'] is None
+            and 'dy' in bound_args.arguments and bound_args.arguments['dy'] is None
+        ):
+            if grid_prototype is not None:
+                grid_deltas = grid_prototype.metpy.grid_deltas
+                bound_args.arguments['dx'] = grid_deltas['dx']
+                bound_args.arguments['dy'] = grid_deltas['dy']
+            elif longitude is not None and latitude is not None and crs is not None:
+                # TODO: de-duplicate .metpy.grid_deltas code
+                bound_args.arguments['dx'], bound_args.arguments['dy'] = (
+                    nominal_lat_lon_grid_deltas(longitude, latitude, crs.get_geod())
+                )
```

Comment:
I've added that now, though not in this block since it's not the xarray case (I'm not interested in having someone pass in only lat/lon manually, but we provide the CRS). I adjusted `grid_deltas` for that case.